File: stripe/src/test/java/com/stripe/android/model/CustomerSourceTest.java
package com.stripe.android.model;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;

import static com.stripe.android.model.CardTest.JSON_CARD_USD;
import static com.stripe.android.model.SourceTest.EXAMPLE_JSON_SOURCE_WITHOUT_NULLS;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

/**
 * Test class for {@link CustomerSource} model class.
 */
public class CustomerSourceTest {

    @Test
    public void fromJson_whenCard_createsCustomerSourceData() throws JSONException {
        final JSONObject jsonCard = new JSONObject(JSON_CARD_USD);
        final CustomerSource sourceData = CustomerSource.fromJson(jsonCard);
        assertNotNull(sourceData);
        assertNotNull(sourceData.asCard());
        assertEquals("card_189fi32eZvKYlo2CHK8NPRME", sourceData.getId());
        assertNull(sourceData.getTokenizationMethod());
    }

    @Test
    public void fromJson_whenCardWithTokenization_createsSourceDataWithTokenization()
            throws JSONException {
        final JSONObject jsonCard = SourceFixtures.APPLE_PAY;
        final CustomerSource sourceData = CustomerSource.fromJson(jsonCard);
        assertNotNull(sourceData);
        assertNotNull(sourceData.asCard());
        assertEquals("card_189fi32eZvKYlo2CHK8NPRME", sourceData.getId());
        assertEquals("apple_pay", sourceData.getTokenizationMethod());
    }

    @Test
    public void fromJson_whenSource_createsCustomerSourceData() {
        final CustomerSource sourceData =
                CustomerSource.fromString(EXAMPLE_JSON_SOURCE_WITHOUT_NULLS);
        assertNotNull(sourceData);
        assertNotNull(sourceData.asSource());
        assertEquals("src_19t3xKBZqEXluyI4uz2dxAfQ", sourceData.getId());
    }

    @Test
    public void fromExampleJsonSource_toJson_createsSameObject() {
        assertNotNull(CustomerSource.fromString(EXAMPLE_JSON_SOURCE_WITHOUT_NULLS));
    }

    @Test
    public void getSourceType_whenCard_returnsCard() throws JSONException {
        final CustomerSource sourceData = CustomerSource.fromJson(new JSONObject(JSON_CARD_USD));
        assertNotNull(sourceData);
        assertEquals(Source.SourceType.CARD, sourceData.getSourceType());
    }

    @Test
    public void getSourceType_whenSourceThatIsNotCard_returnsSourceType() {
        final CustomerSource alipaySource = CustomerSource.fromJson(SourceFixtures.ALIPAY_JSON);
        assertNotNull(alipaySource);
        assertEquals(Source.SourceType.ALIPAY, alipaySource.getSourceType());
    }
}

File: stripe/src/test/java/com/stripe/android/model/CustomerSourceTest.kt
package com.stripe.android.model

import com.stripe.android.model.SourceFixtures.CUSTOMER_SOURCE_CARD
import com.stripe.android.model.parsers.CustomerSourceJsonParser
import org.json.JSONObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

/**
 * Test class for [CustomerSource] model class.
 */
class CustomerSourceTest {

    @Test
    fun fromJson_whenCard_createsCustomerSourceData() {
        val sourceData = requireNotNull(parse(CardFixtures.CARD_USD_JSON))
        assertNotNull(sourceData.asCard())
        assertEquals("card_189fi32eZvKYlo2CHK8NPRME", sourceData.id)
        assertNull(sourceData.tokenizationMethod)
    }

    @Test
    fun fromJson_whenCardWithTokenization_createsSourceDataWithTokenization() {
        val sourceData = requireNotNull(parse(SourceFixtures.APPLE_PAY))
        assertNotNull(sourceData.asCard())
        assertEquals("card_189fi32eZvKYlo2CHK8NPRME", sourceData.id)
        assertEquals(TokenizationMethod.ApplePay, sourceData.tokenizationMethod)
    }

    @Test
    fun fromJson_whenSource_createsCustomerSourceData() {
        assertNotNull(CUSTOMER_SOURCE_CARD.asSource())
        assertEquals("src_19t3xKBZqEXluyI4uz2dxAfQ", CUSTOMER_SOURCE_CARD.id)
    }

    @Test
    fun fromExampleJsonSource_toJson_createsSameObject() {
        assertNotNull(CUSTOMER_SOURCE_CARD)
    }

    @Test
    fun getSourceType_whenCard_returnsCard() {
        val sourceData = parse(CardFixtures.CARD_USD_JSON)
        assertNotNull(sourceData)
        assertEquals(Source.SourceType.CARD, sourceData.sourceType)
    }

    @Test
    fun getSourceType_whenSourceThatIsNotCard_returnsSourceType() {
        val alipaySource = parse(SourceFixtures.ALIPAY_JSON)
        assertNotNull(alipaySource)
        assertEquals(Source.SourceType.ALIPAY, alipaySource.sourceType)
    }

    private companion object {
        private fun parse(jsonObject: JSONObject): CustomerSource? {
            return CustomerSourceJsonParser().parse(jsonObject)
        }
    }
}
